content/algorithms/chain_exchange.md

```diff
@@ -0,0 +1,116 @@
+---
+title: "ChainExchange"
+weight: 5
+dashboardWeight: 1
+dashboardState: stable
+dashboardAudit: missing
+dashboardTests: 0
+---
+
+# ChainExchange
+
+{{< hint info >}}
+**Name**: Chain Exchange
+**Protocol ID**: `/fil/chain/xchg/0.0.1`
+{{< /hint >}}
+
+ChainExchange is a simple request/response protocol that allows Filecoin nodes to request ranges of Tipsets and/or Messages from each other.
+
+The `Request` message requests a chain segment of a given length by the hash of the highest Tipset in the segment (not necessarily heaviest Tipset of the current chain). For example, if the current height is at 5000, but a node is missing Tipsets between 4500-4700, then the `Request.Head` requested is 4700 and `Request.Length` is 200.
+
+The `Options` allow the requester to specify whether they want to receive block headers of the Tipsets only, the transaction messages included in every block, or both.
+
+The `Response` contains the requested chain segment in reverse iteration order. Each item in the `Chain` array contains either the block headers for that Tipset if the `Blocks` option bit in the request is set, or the messages across all blocks in that Tipset, if the `Messages` bit is set, or both, if both option bits are set.
+
+Each `CompactedMessages` structure contains the BLS and `secp256k1` messages for the corresponding Tipset in unified arrays that encode each message once regardless of how many times it appears in the Tipset's blocks (to reduce the `Response` size). The mapping between message and the blocks into which they are included is encoded in the `BlsIncludes` and `SecpkIncludes` arrays. These arrays are indexed by the block index in the Tipset and contain in that position an array of message indexes that belong to that block. The messages themselves are in the `Bls` and `Secpk` arrays.
+
+If not all Tipsets requested could be fetched, but the `Head` of the chain segment requested was successfully fetched (and potentially more contiguous Tipsets), then this is not considered an error and a `Partial` response code is returned. The node can continue extending the chain from the partial returned segment onwards.
+
+```go
+type Request struct {
+    ## Head of the requested segment (block CIDs comprising the entire Tipset)
+	Head [Cid]
+    ## Length of the requested segment
+	Length UInt
+    ## Query options
+    Options UInt
+}
+```
+
+```go
+type Options enum {
+    # Include block headers
+    | Headers 1
+    # Include block messages
+    | Messages 2
+}
+
+type Response struct {
+    ## Response Status
+    Status status
+    ## Optional error message
+    ErrorMessage string
+    ## Returned segment containing block messages and/or headers
+    Chain []*BSTipSet
```

Comment:
You didn't include the BSTipSet type in the spec, and not sure if intentional.

I only mention because each "BSTipSet", or whatever you want to name it, has optional `CompactedMessages` and that isn't noted anywhere

Side note: may want to change the name of this type in the spec? (BS -> BlockSync)

Reply:
Thanks, makes sense. Yes, it's inherited from the previous implementation.

@yiannisbot Posting here Lotus definition of the structures for you to include (if you think useful) in your review pass:
```go
// FIXME: Rename.
type BSTipSet struct {
	// List of blocks belonging to a single tipset to which the
	// `CompactedMessages` are linked.
	Blocks   []*types.BlockHeader
	Messages *CompactedMessages
}

// All messages of a single tipset compacted together instead
// of grouped by block to save space, since there are normally
// many repeated messages per tipset in different blocks.
//
// `BlsIncludes`/`SecpkIncludes` matches `Bls`/`Secpk` messages
// to blocks in the tipsets with the format:
// `BlsIncludes[BI][MI]`
//  * BI: block index in the tipset.
//  * MI: message index in `Bls` list
//
// FIXME: The logic to decompress this structure should belong
//  to itself, not to the consumer.
type CompactedMessages struct {
	Bls         []*types.Message
	BlsIncludes [][]uint64

	Secpk         []*types.SignedMessage
	SecpkIncludes [][]uint64
}
```

Reply:
It is, there are lot of things that still need to be renamed.